component/sqs/component.go

```diff
@@ -0,0 +1,332 @@
+)
+
+// ProcessorFunc definition of a async processor.
+type ProcessorFunc func(context.Context, Batch)
+
+type messageState string
+
+const (
+	sqsAttributeApproximateNumberOfMessages           = "ApproximateNumberOfMessages"
+	sqsAttributeApproximateNumberOfMessagesDelayed    = "ApproximateNumberOfMessagesDelayed"
+	sqsAttributeApproximateNumberOfMessagesNotVisible = "ApproximateNumberOfMessagesNotVisible"
+	sqsAttributeSentTimestamp                         = "SentTimestamp"
+
+	sqsMessageAttributeAll = "All"
+
+	consumerComponent = "sqs-consumer"
+
+	ackMessageState     messageState = "ACK"
+	nackMessageState    messageState = "NACK"
+	fetchedMessageState messageState = "FETCHED"
+)
+
+var (
+	messageAge     *prometheus.GaugeVec
+	messageCounter *prometheus.CounterVec
+	queueSize      *prometheus.GaugeVec
+)
+
+func init() {
+	messageAge = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: "component",
+			Subsystem: "sqs",
+			Name:      "message_age",
+			Help:      "Message age based on the SentTimestamp SQS attribute",
+		},
+		[]string{"sqsAPI"},
+	)
+	prometheus.MustRegister(messageAge)
+	messageCounter = prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Namespace: "component",
+			Subsystem: "sqs",
+			Name:      "message_counter",
+			Help:      "Message counter",
+		},
+		[]string{"sqsAPI", "state", "hasError"},
+	)
+	prometheus.MustRegister(messageCounter)
+	queueSize = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: "component",
+			Subsystem: "sqs",
+			Name:      "queue_size",
+			Help:      "Queue size reported by AWS",
+		},
+		[]string{"state"},
+	)
+	prometheus.MustRegister(queueSize)
+}
+
+// Component implementation of a async component.
+type Component struct {
+	name              string
+	queueName         string
+	queueURL          string
+	sqsAPI            sqsiface.SQSAPI
+	maxMessages       *int64
+	pollWaitSeconds   *int64
+	visibilityTimeout *int64
+	statsInterval     time.Duration
+	proc              ProcessorFunc
+	retries           uint
+	retryWait         time.Duration
+}
+
+// New creates a new component with support for functional configuration.
+func New(name, queueName string, sqsAPI sqsiface.SQSAPI, proc ProcessorFunc, oo ...OptionFunc) (*Component, error) {
+	if name == "" {
+		return nil, errors.New("component name is empty")
+	}
+
+	if queueName == "" {
+		return nil, errors.New("queue name is empty")
+	}
+
+	if sqsAPI == nil {
+		return nil, errors.New("SQS API is nil")
+	}
+
+	if proc == nil {
+		return nil, errors.New("process function is nil")
+	}
+
+	out, err := sqsAPI.GetQueueUrlWithContext(context.Background(), &sqs.GetQueueUrlInput{
+		QueueName: aws.String(queueName),
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to get queue URL: %w", err)
+	}
+
+	cmp := &Component{
+		name:              name,
+		queueName:         queueName,
+		queueURL:          aws.StringValue(out.QueueUrl),
+		sqsAPI:            sqsAPI,
```

Comment:
would it make sense to use constants for the defaults ?

Reply:
done